tsnet/tsnet_test.go#TestConn test flake

### What is the issue?

[tsnet.TestConn](https://github.com/tailscale/tailscale/blob/v1.69.0-pre/tsnet/tsnet_test.go#L230) flaked on an urelated PR https://github.com/tailscale/tailscale/actions/runs/9872700293/job/27263341563?pr=12765#step:8:299:

```
...
=== RUN   TestConn
    tsnet_test.go:111: DERP httpsrv listener: 127.0.0.1:36087
    tsnet_test.go:123: testcontrol listening on http://127.0.0.1:41905/
2024/07/10 10:40:53 tsnet starting with hostname "s1", varRoot "/tmp/TestConn1697026267/001/s1"
2024/07/10 10:40:54 LocalBackend state is NeedsLogin; running StartLoginInteractive...
2024/07/10 10:40:54 tsnet starting with hostname "s2", varRoot "/tmp/TestConn1697026267/002/s2"
2024/07/10 10:40:55 LocalBackend state is NeedsLogin; running StartLoginInteractive...
2024/07/10 10:40:59 AuthLoop: state is Running; done
2024/07/10 10:41:00 AuthLoop: state is Running; done
    tsnet_test.go:255: error Post "http://local-tailscaled.sock/localapi/v0/ping?ip=100.64.0.1&size=0&type=ICMP": context deadline exceeded: 
    stuntest.go:65: STUN server shutdown
--- FAIL: TestConn (30.01s)
...
```

### Steps to reproduce

_No response_

### Are there any recent changes that introduced the issue?

_No response_

### OS

_No response_

### OS version

_No response_

### Tailscale version

_No response_

### Other software

_No response_

### Bug report

_No response_

I’m seeing this as well:
```
--- FAIL: TestPingPeerLearnedViaDelta (66.22s)
    --- FAIL: TestPingPeerLearnedViaDelta/disco (60.54s)
        tsnet_test.go:211: DERP httpsrv listener: 127.0.0.1:33929
        tsnet_test.go:224: testcontrol listening on http://127.0.0.1:41355
        testcontrol.go:369: Created user userid:1 for node nodekey:772803eb2a1f0da181a9ae29a2159b7d01252645b8a7e9017ba0667ba0a1b914
        testcontrol.go:369: Created user userid:2 for node nodekey:0e11afcf861bd12b0e4ceee0995424761024c9314f62b28b2922eeb60d96ed04
        tsnet_test.go:1805: Ping(disco): error Post "http://local-tailscaled.sock/localapi/v0/ping?ip=100.64.0.2&size=0&type=disco": context deadline exceeded: 
        stuntest.go:65: STUN server shutdown
FAIL
FAIL	tailscale.com/tsnet	126.042s
```

--- a/tsnet/tsnet_test.go
+++ b/tsnet/tsnet_test.go
@@ -65,6 +65,15 @@ import (
 	"tailscale.com/util/must"
 )
 
+// pingTimeout returns a per-ping budget for use within the larger test ctx:
+// enough headroom for wireguard-go's RekeyTimeout (5s) plus the actual handshake on slow CI
+// (notably GOARCH=386 emulation where Curve25519/ChaCha20 lack the amd64 assembly fast paths),
+// but tight enough that a hung Ping points the finger at its callsite
+// rather than swallowing the whole test deadline.
+func pingTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
+	return context.WithTimeout(ctx, 10*time.Second)
+}
+
 // TestListener_Server ensures that the listener type always keeps the Server
 // method, which is used by some external applications to identify a tsnet.Listener
 // from other net.Listeners, as well as access the underlying Server.
@@ -336,7 +345,35 @@ func startServer(t *testing.T, ctx context.Context, controlURL, hostname string)
 	}
 	s.lb.ConfigureCertsForTest(testCertRoot.getCert)
 
+	// Wait for the server to finish connecting to its home DERP server,
+	// to prevent fast tests from racing the DERP handshake resulting
+	// in a dropped request with PeerGoneNotHere.
+	waitForHomeDERPConnected(t, ctx, s)
+
 	return s, status.TailscaleIPs[0], status.Self.PublicKey
+}
+
+// waitForHomeDERPConnected blocks until s has selected a home DERP region
+// and received its first frame from that region.
+// Until s establishes a complete connection to its home DERP server,
+// the DERP server will drop any incoming peer DISCO frames looking for s
+// with PeerGoneNotHere.
+func waitForHomeDERPConnected(t testing.TB, ctx context.Context, s *Server) {
+	t.Helper()
+	h := s.Sys().HealthTracker.Get()
+	ms := s.Sys().MagicSock.Get()
+	for {
+		if r := ms.GetLastNetcheckReport(ctx); r != nil &&
+			r.PreferredDERP != 0 &&
+			!h.GetDERPRegionReceivedTime(r.PreferredDERP).IsZero() {
+			return
+		}
+		select {
+		case <-ctx.Done():
+			t.Fatalf("waitForHomeDERPConnected(%s): %v", s.hostname, ctx.Err())
+		case <-time.After(20 * time.Millisecond):
+		}
+	}
 }
 
 func TestDialBlocks(t *testing.T) {
@@ -427,7 +464,9 @@ func TestConn(t *testing.T) {
 	}))
 
 	// ping to make sure the connection is up.
-	res, err := lc2.Ping(ctx, s1ip, tailcfg.PingTSMP)
+	pingCtx, cancelPing := pingTimeout(ctx)
+	defer cancelPing()
+	res, err := lc2.Ping(pingCtx, s1ip, tailcfg.PingTSMP)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -1668,7 +1707,9 @@ func TestFallbackTCPHandler(t *testing.T) {
 	}
 
 	// ping to make sure the connection is up.
-	res, err := lc2.Ping(ctx, s1ip, tailcfg.PingICMP)
+	pingCtx, cancelPing := pingTimeout(ctx)
+	defer cancelPing()
+	res, err := lc2.Ping(pingCtx, s1ip, tailcfg.PingICMP)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -1792,13 +1833,7 @@ func testPingPeerLearnedViaDelta(t *testing.T, pt tailcfg.PingType) {
 		t.Fatal(err)
 	}
 
-	// Per-ping budget within the larger test ctx: enough headroom for
-	// wireguard-go's RekeyTimeout (5s) plus the actual handshake on
-	// slow CI (notably GOARCH=386 emulation where Curve25519/ChaCha20
-	// lack the amd64 assembly fast paths), but tight enough that a
-	// hung Ping points the finger at this call rather than swallowing
-	// the whole test deadline.
-	pingCtx, cancelPing := context.WithTimeout(ctx, 60*time.Second)
+	pingCtx, cancelPing := pingTimeout(ctx)
 	defer cancelPing()
 	pr, err := lc1.Ping(pingCtx, s2ip, pt)
 	if err != nil {
@@ -1897,7 +1932,7 @@ func TestPingSubnetRouteOfDeltaPeer(t *testing.T) {
 	// the bug: a stale BART / lastCfgFull (PeerForIP / lookupPeerByIP
 	// miss) AND a stale wgdev PeerLookupFunc closure (peer's noise
 	// key not yet registered for outbound encryption).
-	pingCtx, cancelPing := context.WithTimeout(ctx, 60*time.Second)
+	pingCtx, cancelPing := pingTimeout(ctx)
 	defer cancelPing()
 	pr, err := lc1.Ping(pingCtx, probeIP, tailcfg.PingTSMP)
 	if err != nil {
@@ -1928,7 +1963,9 @@ func TestPingSelfReturnsIsLocalIP(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	pr, err := lc.Ping(ctx, s1ip, tailcfg.PingDisco)
+	pingCtx, cancelPing := pingTimeout(ctx)
+	defer cancelPing()
+	pr, err := lc.Ping(pingCtx, s1ip, tailcfg.PingDisco)
 	if err != nil {
 		t.Fatalf("Ping: %v", err)
 	}
@@ -1961,7 +1998,9 @@ func TestCapturePcap(t *testing.T) {
 	}
 
 	// send a packet which both nodes will capture
-	res, err := lc2.Ping(ctx, s1ip, tailcfg.PingICMP)
+	pingCtx, cancelPing := pingTimeout(ctx)
+	defer cancelPing()
+	res, err := lc2.Ping(pingCtx, s1ip, tailcfg.PingICMP)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -2008,7 +2047,9 @@ func TestUDPConn(t *testing.T) {
 	}
 
 	// ping to make sure the connection is up.
-	res, err := lc2.Ping(ctx, s1ip, tailcfg.PingICMP)
+	pingCtx, cancelPing := pingTimeout(ctx)
+	defer cancelPing()
+	res, err := lc2.Ping(pingCtx, s1ip, tailcfg.PingICMP)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -2226,7 +2267,9 @@ func TestUserMetricsByteCounters(t *testing.T) {
 	})
 
 	// ping to make sure the connection is up.
-	res, err := lc2.Ping(ctx, s1ip, tailcfg.PingICMP)
+	pingCtx, cancelPing := pingTimeout(ctx)
+	defer cancelPing()
+	res, err := lc2.Ping(pingCtx, s1ip, tailcfg.PingICMP)
 	if err != nil {
 		t.Fatalf("pinging: %s", err)
 	}
@@ -2608,7 +2651,10 @@ func setupTwoClientTest(t *testing.T, useTUN bool) *listenTest {
 	waitForPeerReachable(t, s2, s1.lb.NodeKey())
 
 	lc1 := must.Get(s1.LocalClient())
-	must.Get(lc1.Ping(ctx, s2ip4, tailcfg.PingTSMP))
+
+	pingCtx, cancelPing := pingTimeout(ctx)
+	defer cancelPing()
+	must.Get(lc1.Ping(pingCtx, s2ip4, tailcfg.PingTSMP))
 
 	return &listenTest{
 		control: control,

